datastore/src/main/java/com/google/datastore/snippets/Concepts.java

```diff
@@ -0,0 +1,1058 @@
+    Entity task = Entity.builder(taskKey)
+        .set("tags", StringValue.of("fun"), StringValue.of("programming"))
+        .set("collaborators", StringValue.of("alice"), StringValue.of("bob"))
+        .build();
+    // [END array_value]
+    assertValidEntity(task);
+  }
+
+  @Test
+  public void testBasicEntity() {
+    // [START basic_entity]
+    Entity task = Entity.builder(taskKey)
+        .set("type", "Personal")
+        .set("done", false)
+        .set("priority", 4)
+        .set("description", "Learn Cloud Datastore")
+        .build();
+    // [END basic_entity]
+    assertValidEntity(task);
+  }
+
+  @Test
+  public void testUpsert() {
+    // [START upsert]
+    Key taskKey = datastore.allocateId(keyFactory.newKey());
+    Entity task = Entity.builder(taskKey).build();
+    datastore.put(task);
+    // [END upsert]
+    assertEquals(task, datastore.get(task.key()));
+  }
+
+  @Test
+  public void testInsert() {
+    // [START insert]
+    Entity task = Entity.builder(taskKey).build();
+    Key taskKey = datastore.add(task).key();
+    // [END insert]
+    assertEquals(task, datastore.get(taskKey));
+  }
+
+  @Test
+  public void testLookup() {
+    datastore.put(testEntity);
+    // [START lookup]
+    Entity task = datastore.get(taskKey);
+    // [END lookup]
+    assertEquals(testEntity, task);
+  }
+
+  @Test
+  public void testUpdate() {
+    datastore.put(testEntity);
+    // [START update]
+    Entity task = Entity.builder(datastore.get(taskKey)).set("priority", 5).build();
+    datastore.update(task);
+    // [END update]
+    assertEquals(task, datastore.get(taskKey));
+  }
+
+  @Test
+  public void testDelete() {
+    datastore.put(testEntity);
+    // [START delete]
+    datastore.delete(taskKey);
+    // [END delete]
+    assertNull(datastore.get(taskKey));
+  }
+
+  private List<Entity> setUpBatchTests(Key taskKey1, Key taskKey2) {
+    Entity task1 = Entity.builder(taskKey1)
+        .set("type", "Personal")
+        .set("done", false)
+        .set("priority", 4)
+        .set("description", "Learn Cloud Datastore")
+        .build();
+    Entity task2 = Entity.builder(taskKey2)
+        .set("type", "Personal")
+        .set("done", false)
+        .set("priority", 5)
+        .set("description", "Integrate Cloud Datastore")
+        .build();
+    datastore.put(task1, task2);
+    return ImmutableList.of(task1, task2);
+  }
+
+  @Test
+  public void testBatchUpsert() {
+    Key taskKey1 = keyFactory.newKey(1);
+    Key taskKey2 = keyFactory.newKey(2);
+    Entity task1 = Entity.builder(taskKey1)
+        .set("type", "Personal")
+        .set("done", false)
+        .set("priority", 4)
+        .set("description", "Learn Cloud Datastore")
+        .build();
+    Entity task2 = Entity.builder(taskKey2)
+        .set("type", "Personal")
+        .set("done", false)
+        .set("priority", 5)
+        .set("description", "Integrate Cloud Datastore")
+        .build();
+    // [START batch_upsert]
```

Comment:
Since v1beta2 doesn't support upsert with auto id allocation, can we make this a batch insert instead of upsert? Sorry I didn't catch this, these snippets were designed for v1beta3.